samigo/samigo-impl/src/java/org/sakaiproject/samigo/impl/SamigoAvailableNotificationServiceImpl.java

```diff
@@ -0,0 +1,369 @@
+                log.debug("Samigo assessment site is published and is not softly deleted.");
+                log.debug("Release to: {}", publishedAssessment.getAssessmentAccessControl().getReleaseTo());
+                log.debug("Execution has reached the inside of the main published/not deleted/not dead/Samigo-exists block.");
+                if (!StringUtils.isEmpty(extensionId)){ //first, we need to deal with the possibility that this could be an Exception email.
+                    ExtendedTime extension = PersistenceService.getInstance().getExtendedTimeFacade().getEntry(extensionId);
+                    if(extension != null){  //make sure the extension exists
+                        if (!StringUtils.isEmpty(extension.getUser())){ //when the extension has an individual user
+                            sendEmailNotification(site, publishedAssessment, site.getMember(extension.getUser()), extension);
+                        }
+                        if (!StringUtils.isEmpty(extension.getGroup())){    //extension for a group
+                            for(Member member: site.getGroup(extension.getGroup()).getMembers()){
+                                sendEmailNotification(site, publishedAssessment, member, extension);
+                            }
+                        }
+                    }
+                } else if (StringUtils.equals(publishedAssessment.getAssessmentAccessControl().getReleaseTo(), "Selected Groups")){ //when there is a releaseTo setting that limits the access of the test
+                    for (Object groupId : publishedAssessment.getReleaseToGroups().keySet().toArray()){ //loop through applicable group IDs
+                        for (Member member : site.getGroup((String) groupId).getMembers()){  //loop through users of applicable group.
+                            if (member.isActive() && !isUserInException(publishedAssessment, member.getUserId(), site)) {
+                                log.debug("Calling send email notification: {}", member.getUserId());
+                                sendEmailNotification(site, publishedAssessment, member, null);
+                            }
+                        }
+                    }
+                } else if(StringUtils.equals(publishedAssessment.getAssessmentAccessControl().getReleaseTo(), "Anonymous Users")){
+                    //don't do anything when we've released to Anonymous Users.
+                } else {    //if there is no special ReleaseTo setting, we can email all site members.
+                    for (Member member : site.getMembers()) {
+                        log.debug("Samigo assessment open notification checking if member needs an email, member: {}", member.getUserId());
+                        log.debug("Member info: active? {}", member.isActive());
+                        log.debug("Member info: released to this member? {}", "TODO: THIS");
```

Comment:
This seems a little odd. Is this something that still needs to be done, denoted by "TODO"? If not, there is no string replacement necessary, just combine them:
```suggestion
                        log.debug("Member info: released to this member? TODO: THIS");
```